packit/schema.py

```diff
@@ -519,22 +537,63 @@ def rename_deprecated_keys(self, data, **kwargs):
         return data
 
     @staticmethod
-    def propagate_options_to_jobs(data, **_):
-        """
-        add all the fields (except for jobs) to every job so we can process only jobconfig in p-s
-        """
-        if not data:  # data is None when .packit.yaml is empty
-            return data
-        for job in data.get("jobs", []):
-            for k, v in data.items():
-                if k == "jobs":
-                    # overriding jobs doesn't make any sense
-                    continue
-                job.setdefault(k, v)
+    def rearrange_packages(data: dict) -> dict:
+        packages = data.pop("packages")
+        jobs = data.pop("jobs")
+        for k, v in packages.items():
+            # First set the defaults which are not inherited from
+            # the top-level, in case they are not set yet.
+            v.setdefault("downstream_package_name", k)
+            # Inherit default values from the top-level.
+            v.update(data | v)
+        data = {"packages": packages, "jobs": jobs}
+        return data
+
+    @staticmethod
+    def rearrange_jobs(data: dict) -> dict:
+        packages = data.pop("packages")
+        jobs = data.pop("jobs")
+        for i, job in enumerate(jobs):
+            job_type = job.pop("job")
+            trigger = job.pop("trigger")
+            selected_packages = job.pop("packages", None)
+            # There is no 'packages' key in the job, so
+            # the job should handle all the top-level packages.
+            if not selected_packages:
+                jobs[i] = {
+                    "job": job_type,
+                    "trigger": trigger,
+                    "packages": {k: v | job for k, v in packages.items()},
+                }
+            # Some top-level packages are selected to be
+            # handled by the job.
+            elif isinstance(selected_packages, list):
+                jobs[i] = {
+                    "job": job_type,
+                    "trigger": trigger,
+                    "packages": {
+                        k: v | job
+                        for k, v in packages.items()
+                        if k in selected_packages
+                    },
+                }
+            # Some top-level packages are selected to be
+            # handled by the job AND have some custom config.
+            elif isinstance(selected_packages, dict):
+                jobs[i] = {
+                    "job": job_type,
+                    "trigger": trigger,
+                    "packages": {
+                        k: packages[k] | job | v for k, v in selected_packages.items()
+                    },
+                }
+            else:
+                raise ValidationError("unknown type")
```

Comment:
could you please improve this error message so it's clear that it's about the 'packages' key?

Reply:
Reworked this part, should be fixed now.